### DISCUSSION TITLE
Console errors running Keyboard Maestro macros via AppleScript and Bunch

### DISCUSSION THREAD

**Discussion by rbanks88** [ttscoff/bunch-docs#186] [Q&A]
I'm new to Bunch, and seem to have hit a wall already. I'm trying to run a Keyboard Maestro macro via an AppleScript that a Bunch runs, but the macro doesn't run and I'm getting errors in the Console. Any idea what I'm doing wrong?

Here's the situation:

I have a uniquely named Keyboard Maestro macro, "Show OmniFocus Inspector".

This AppleScript, named Show_OmniFocus_Inspector, in the Bunches folder, should run that macro:

tell application "Keyboard Maestro Engine"
  do script "Show OmniFocus Inspector"
end tell

I run the AppleScript with this line in my bunch:

* Show_OmniFocus_Inspector.scpt // runs applescript in Bunches that runs Keyboard Maestro macro

I get these errors in the Console:

2022-01-12 18:51:41: 🔴Script Error: exit code 1 for /usr/bin/osascript /Users/rcb/Documents/Computing/Bunches/Show_OmniFocus_Inspector.scpt
2022-01-12 18:51:41: 🔴STDERR: /Users/rcb/Documents/Computing/Bunches/Show_OmniFocus_Inspector.scpt:41:94: execution error: Keyboard Maestro Engine got an error: do script found no macros with a matching name (macros must be enabled, and in macro groups that are enabled and currently active). (-1)

**Comment by ttscoff** [comment]
Does the script succeed when run from within Script Editor?

-Brett

**Reply by rbanks88** [reply]
It didn't run from the script editor, but the error gave clues and I then realized I needed to move the Keyboard Maestro macro out of the OmniFocus folder into a global folder, since OmniFocus wasn't "front" when the macro was called. I'm getting no Console errors now, but even though "Launch items in Parallel" is unchecked, Bunch execution appears to be jumping around in the script.
 
Based on what I see on the screen, it's executing the steps in the order I've numbered on the attachment. Do I need to put pauses on items so they execute in the desired order? Thanks! 

![Scan Jan 13, 2022, 10 42, 02- rename](https://user-images.githubusercontent.com/10098078/149381781-d5ee8b4f-5b4e-4088-8be4-66fb53a440b2.jpg)

**Comment by ttscoff** [comment]
When not running parallel, commands are sent in order, but still very 
rapidly. So your apps are launching in order, but the speed at which 
they respond to the series of commands sent to them determines the order 
in which they're actually processed. A window receiving a bunch of 
keystrokes all at once may queue and process them in an unexpected 
order.

I would recommend combining all of the keyboard sequences where 
possible. (and ~1 doesn't send a pause, that sends Option-1. To include 
a pause in a keyboard sequence, use an exclamation point: !1.)

For example:

```
BusyCal
- ***@***.*** !1 @2 !1 @t (left)3}
* script...

OmniFocus
- ***@***.*** @1 !1}
* scripts...
- (pause 3) // give the scripts a chance to execute
- {@~n}
```

Let me know if that helps.

**Reply by rbanks88** [reply]
That was a big help. I got the idea that ~ meant pause from the Quick Reference page that said App Name ~5 meant Launch app after 5 second delay. I didn't catch the fact that you delay with ~ for apps and ! for commands.

It's still executing in an order I can't fathom, though. The BusyCal section is fine, but OmniFocus appears to launch with the window(s) open when it was last quit, then it skips the first Bunch line and runs the AppleScript/Keyboard Maestro macros (I put a "speak text" action in the last Keyboard Maestro macro). Then it executes the first bunch line, closing all windows and opening the Inbox. And it never executes the final Bunch lines (Cmd-Opt N to open a new window), even though I put it in twice. Here's the code:

`---
title: Morning Startup
---

# Launch BusyCal in week view with today in the middle
BusyCal
- {@0 !1 @2 !1 @t (left)3}

OmniFocus
- {~@w !1 @1} // Close all windows via Opt Cmd w, Open inbox with 1 sec. pause after
- (pause 3)
* Hide_OmniFocus_Sidebar.scpt // runs applescript in Bunches that runs Keyboard Maestro macro
* Show_OmniFocus_Inspector.scpt // runs applescript in Bunches that runs Keyboard Maestro macro
- (pause 3) // give the scripts a chance to execute
- {@~n} // Open new window via Opt Cmd n with 1 sec. pause after
- {@~n} // Open new window via Opt Cmd n with 1 sec. pause after
`

I have this all working in a long Keyboard Maestro macro, but thought it would be more straightforward, and more flexible, in Bunch. However, I'm hitting so many roadblocks... I'm an artist, not an IT guy... so this may be too far out of my wheelhouse!

Best,
Russell

**Reply by ttscoff** [reply]
I just figured out what's probably going on, not sure why I didn't see it immediately. Probably because the asterisks and dashes were all just rendering as a bullet list so I didn't think about it... 

When you switch to AppleScript lines (asterisk) you're no longer sending anything directly to OmniFocus. Those are executing in the background, and when you go back to the keystroke lines, Bunch no longer knows where to send them, as it's dissociated from the original app line (OmniFocus). You can always repeat an app to send additional commands to it, or focus it, or anything necessary, it will only launch once. What you need to do is break it up into multiple parts:

```
OmniFocus
- {~@w !1 @1} // Close all windows via Opt Cmd w, Open inbox with 1 sec. pause after

* Hide_OmniFocus_Sidebar.scpt // runs applescript in Bunches that runs Keyboard Maestro macro
* Show_OmniFocus_Inspector.scpt // runs applescript in Bunches that runs Keyboard Maestro macro

OmniFocus
- (pause 3) // give the scripts a chance to execute
- {@~n} // Open new window via Opt Cmd n with 1 sec. pause after
```

Also, is there a reason you need two separate scripts? Can't they be combined into one? That would guarantee the execution order as a script will always execute sequentially, where as two scripts run the risk of executing simultaneously and the system would decide which one gets executed first.

(BTW, when you're posting a code block like above, use triple backticks above and below it to render it correctly. A single backtick breaks when there's a newline (code span vs block).)

**Reply by rbanks88** [reply]
Thank you so much for your kind patience! You are a national treasure. 

 It works as desired now, and I now understand why. 
I also combined the two scripts They were running two Keyboard Maestro macros, so I combined the two macros and made one script to trigger it.

And thanks for the tip on the backticks. I was just clicking the "Insert Code" icon in the Github message editor, and thought it looked weird!

Best,
Russell
